src/Plugin/formatter/FormatterHalJson.php

```diff
@@ -186,6 +190,21 @@ protected function extractFieldValues($rows) {
       if (!$rows instanceof ResourceFieldCollectionInterface) {
         throw new InternalServerErrorException('Inconsistent output.');
       }
+
+      // This feels a bit awkward, but if the result is going to be cached, it
+      // pays off the extra effort of generating the whole resource entity. That
+      // way we can get a different field set with the previously cached entity.
+      // If the entity is not going to be cached, then avoid generating the
+      // field data altogether.
+      $limit_fields = $rows->getLimitFields();
```

Comment:
When caching is enabled and we get a cache MISS we want to generate output for the cache entry for the **whole** entity. That way we can use that cache record independently of the sparse fieldset.

On the other hand, if cache is not enabled we don't want to output for the whole entity, only the bits that we are going to need. For performance reasons.